runtime/drivers/duckdb/connectors.go

```diff
@@ -13,7 +14,33 @@ import (
 	"github.com/rilldata/rill/runtime/pkg/fileutil"
 )
 
+// Ingest data from a source with a timeout
 func (c *connection) Ingest(ctx context.Context, env *connectors.Env, source *connectors.Source) error {
+	cancellableCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	timeout := 30
+	if value, ok := source.Properties["timeout"]; ok {
+		timeout = int(value.(float64))
+	}
+
+	channel := make(chan error, 1)
+	go func() {
+		// relies on duck db query cancellation to cancel the ingestion
+		err := c.ingestWithoutTimeout(cancellableCtx, env, source)
+		channel <- err
+	}()
+
+	select {
+	case result := <-channel:
+		return result
+
+	case <-time.After(time.Duration(timeout) * time.Second):
+		return context.DeadlineExceeded
+	}
```

Comment:
1. In practice, the cancellation will be very fast. Go's networking libraries abort almost instantly on context cancel, and DuckDB will cancel almost instantly when the next release comes (within few weeks). There's a big problem in doing cleanups in the background, which is that the caller might think the DB is not busy (blocked), but it actually is. This can cause weird latencies, especially in context of query priorities, which might mean the next queries get interleaved the background cleanup, and so on (we've had similar problems previously).

It might make sense to do some timeout cleanup in the background at some point, but don't want to until we've a) observed it to be a problem, b) know what kind of problems that cause cancellation to take much more than a few ms (if cleanup time is long, it might signal another problem, which could gradually accumulate background resources over time).

2. The note sounds reasonable

Reply:
Makes sense. Please also suggest default timeout value. I kept it as 30 seconds but could be very short for s3/gcs ingestion